carsus/io/cmfgen/util.py

```diff
@@ -0,0 +1,332 @@
+                    break
+
+        else:
+            n, line = None, None
+
+    if row_number is True:
+        return n
+
+    return line
+
+
+def parse_header(fname, keys, start=0, stop=50):
+    """
+    Parse header information from CMFGEN files.
+
+    Parameters
+    ----------
+    fname : str
+        Path to plain text file.
+    keys : list of str
+        Entries to search.
+    start : int
+        First line to search in (default is 0).
+    stop : int
+        Last line to search in (default is 50).
+
+    Returns
+    -------
+    dict
+        Dictionary containing metadata.
+    """
+    meta = {k.strip('!'):None for k in keys}
+
+    with open_cmfgen_file(fname) as f:
+        for line in itertools.islice(f, start, stop):
+            for k in keys:
+                if k.lower() in line.lower():
+                    meta[k.strip('!')] = line.split()[0]
+
+    return meta
+
+
+def get_seaton_phixs_table(threshold_energy_ryd, sigma_t, beta, s, nu_0=None, n_points=1000):
+    """
+    """
+    energy_grid = np.linspace(0.0, 1.0, n_points, endpoint=False)
+    phixs_table = np.empty((len(energy_grid), 2))
+
+    for i, c in enumerate(energy_grid):
+        energy_div_threshold = 1 + 20 * (c ** 2)
+
+        if nu_0 is None:
+            threshold_div_energy = energy_div_threshold ** -1
+            cross_section = sigma_t * (beta + (1 - beta) * threshold_div_energy) * (threshold_div_energy ** s)
+
+        else:
+            threshold_energy_ev = threshold_energy_ryd * RYD_TO_EV
+            energy_offset_div_threshold = energy_div_threshold + (nu_0 * 1e15 * H_IN_EV_SECONDS) / threshold_energy_ev
+            threshold_div_energy_offset = energy_offset_div_threshold ** -1
+
+            if threshold_div_energy_offset < 1.0:
+                cross_section = sigma_t * (beta + (1 - beta) * (threshold_div_energy_offset)) * \
+                    (threshold_div_energy_offset ** s)
+
+            else:
+                cross_section = 0.0
+
+        phixs_table[i] = energy_div_threshold * threshold_energy_ryd, cross_section
+
+    return phixs_table
+
+
+def get_hydrogenic_n_phixs_table(hyd_gaunt_energy_grid_ryd, hyd_gaunt_factor, threshold_energy_ryd, n):
+    """
+    """
+    energy_grid = hyd_gaunt_energy_grid_ryd[n]
+    phixs_table = np.empty((len(energy_grid), 2))
+    scale_factor = 7.91 / threshold_energy_ryd / n
+
+    for i, energy_ryd in enumerate(energy_grid):
+        energy_div_threshold = energy_ryd / energy_grid[0]
+
+        if energy_div_threshold > 0:
+            cross_section = scale_factor * hyd_gaunt_factor[n][i] / (energy_div_threshold) ** 3
+        else:
+            cross_section = 0.0
+
+        phixs_table[i][0] = energy_div_threshold * threshold_energy_ryd
+        phixs_table[i][1] = cross_section
+
+    return phixs_table
+
+
+def get_hydrogenic_nl_phixs_table(hyd_phixs_energy_grid_ryd, hyd_phixs, threshold_energy_ryd, n, l_start, l_end, nu_0=None):
+    """
+    """
+
+    assert l_start >= 0
```

Comment:
Raise exception?

Reply:
Add descriptive output string

Reply:
I don't know if these assertions are useful, probably not.